tango/integrations/fairscale/training_engine.py

```diff
@@ -0,0 +1,157 @@
+        when ``device_count > 1`` in the :class:`~tango.integrations.torch.TorchTrainStep`.
+
+    For maximum memory savings, we recommend training with AMP enabled and the following
+    :class:`FSDPConfig`:
+
+    .. testcode::
+
+        from tango.integrations.fairscale import FSDPConfig
+
+        fsdp_config = FSDPConfig(
+            reshard_after_forward=True,
+            move_params_to_cpu=True,
+            move_grads_to_cpu=True,
+            mixed_precision=True,
+        )
+
+    For maximum training *speed*, we recommend training with AMP enabled and the following
+    :class:`FSDPConfig`:
+
+    .. testcode::
+
+        from tango.integrations.fairscale import FSDPConfig
+
+        fsdp_config = FSDPConfig(
+            reshard_after_forward=False,
+            move_params_to_cpu=False,
+            move_grads_to_cpu=False,
+            mixed_precision=True,
+        )
+
+    Parameters
+    ----------
+    amp : :class:`bool`, optional
+        Use automatic mixed precision (AMP). Default is ``False``.
+    max_grad_norm : :class:`float`, optional
+        If set, gradients will be clipped to have this max norm. Default is ``None``.
+    amp_use_bfloat16 : ``Optional[bool]``
+        Set to ``True`` to force using the ``bfloat16`` datatype in mixed precision training.
+        Only applicable when ``amp=True``. If not specified, the default behavior will be
+        to use ``bfloat16`` when training with AMP on CPU, otherwise not.
+    fsdp_config : :class:`FSDPConfig`
+        The options for :class:`~fairscale.nn.FullyShardedDataParallel`.
+        If not specified, the default options will be used.
+
+    """
+
+    def __init__(
+        self,
+        train_config: TrainConfig,
+        model: Lazy[Model],
+        optimizer: Lazy[Optimizer],
+        *,
+        lr_scheduler: Optional[Lazy[LRScheduler]] = None,
+        amp: bool = False,
+        max_grad_norm: Optional[float] = None,
+        amp_use_bfloat16: Optional[bool] = None,
+        fsdp_config: Optional[FSDPConfig] = None,
+    ) -> None:
+        if not train_config.is_distributed:
+            raise ConfigurationError(
+                f"{self.__class__.__name__} can only be used with distributed training"
+            )
+
+        self.fsdp_config = fsdp_config or FSDPConfig()
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        super().__init__(
+            train_config,
+            model,
+            optimizer,
+            lr_scheduler=lr_scheduler,
+            amp=amp,
+            max_grad_norm=max_grad_norm,
+            amp_use_bfloat16=amp_use_bfloat16,
+        )
+        if amp:
+            self.grad_scaler = ShardedGradScaler()
+
+    def _construct_model(self, model: Lazy[Model]) -> Model:
+        model: Model = model.construct()
+        if not self.fsdp_config.move_params_to_cpu:
+            model.to(self.train_config.worker_local_default_device)
+        return FSDP(model, **self.fsdp_config.as_kwargs())
```

Comment:
> This does not look like it supports efficient loading of models into GPU memory?

Why not?

Reply:
If you use `with_wrapped_modules()` on the inner modules of your model, it will be efficient. Maybe I forgot to write that in the docs explicitly.

Reply:
e83f972

Reply:
I just see the naked `model.construct()`. At that point, no code that's fairscale specific has run yet. So if `model.construct()` takes 500 GB of memory, we'll need 500 GB of memory on that line.

Reply:
See my previous two comments

Reply:
I did not get this before. But I think now I do. Since the output of `model.construct()` is already wrapped, it's not true that nothing fairscale specific has run. The module wrappers have already been applied, so it's all good.